vertx-core/src/main/java/io/vertx/core/json/JsonArray.java

```diff
@@ -456,7 +456,17 @@ public JsonArray set(int pos, Object value) {
    * @return true if it contains the value, false if not
    */
   public boolean contains(Object value) {
-    return list.contains(value);
+    if (list.contains(value)) {
+      return true;
+    }
+
+    if (value instanceof JsonObject) {
+      return list.contains(((JsonObject) value).getMap());
+    } else if (value instanceof JsonArray) {
+      return list.contains(((JsonArray) value).getList());
+    }
+
+    return false;
```

Comment:
In the worst scenario, we will iterate over the list three times. Besides, we want to be able to check if the JsonArray contains a JsonArray or JsonObject regardless of how it's stored, not to check if a JsonArray contains an arbitrary list or map.

I would suggest to use something like this:

```java
  @SuppressWarnings("unchecked")
  public boolean contains(Object value) {
    if (value == null) {
      for (Object entry : list) {
        if (entry == null) {
          return true;
        }
      }
    } else {
      for (Object entry : list) {
        if (entry instanceof Map && value instanceof JsonObject) {
          return value.equals(new JsonObject((Map<String, Object>) entry));
        }
        if (entry instanceof List && value instanceof JsonArray) {
          return value.equals(new JsonArray((List<String>) entry));
        }
        if (value.equals(entry)) {
          return true;
        }
      }
    }
    return false;
  }
```

Reply:
Thanks for the suggestion @tsegismont. Yes I was debating back and forth whether to essentially re-implement [`ArrayList#contains`](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/util/ArrayList.java#L320) or whether the better code quality with the downside of iterating over the list a maximum number of 2 times (not 3, because it is either a `JsonArray` or `JsonObject` you want to check for) is the better solution. I can agree that we should go for the option that is better from a performance point of view. In my solution I did optimize for doing less comparisons, however I agree that the iteration itself is much more time consuming than doing the comparison operations.

> Besides, we want to be able to check if the `JsonArray` contains a `JsonArray` or `JsonObject` regardless of how it's stored, not to check if a `JsonArray` contains an arbitrary list or map.

Mhm, as far as I am concerned, I would argue that a Map / List inside of a `JsonArray` is virtually equivalent to it's wrapped `JsonArray` / `JsonObject` counterpart. Reason for me is the public interface of `JsonArray` does wrapping on getting anyways, so classic duck-typing principle: If it walks like a duck and it quacks like a duck, then it must be a duck. An arbitrary `Map` / `List` is not supported inside of a `JsonArray`, it is just a "vehicle" for representing nested arrays anyways. This is why for your solution I would argue that creating temporary `JsonObject` / `JsonArray` depending on the size of the `JsonArray` creates a lot of temporary objects and thus a lot of pressure on the heap / garbage collection.

I would maybe suggest to "mix" both solutions and:

a) Keep the assumption that `Map` / `List` inside of `JsonArray` is equivalent to `JsonObject` / `JsonArray` and
b) Re-implement `contains` based on [`ArrayList#contains`](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/util/ArrayList.java#L320) as you suggested.

Let me make a proposal to continue our discussion. 👍 Thanks for the good suggestion.

Reply:
@tsegismont I made a new suggestion in my latest force push.